src/System.Net.Mail/src/System/Net/Mail/SmtpNegotiateAuthenticationModule.cs

```diff
@@ -0,0 +1,199 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Security.Authentication.ExtendedProtection;
+using System.Security.Permissions;
+
+namespace System.Net.Mail
+{
+    internal class SmtpNegotiateAuthenticationModule : ISmtpAuthenticationModule
+    {
+        private Dictionary<object, NTAuthentication> _sessions = new Dictionary<object, NTAuthentication>();
+
+        internal SmtpNegotiateAuthenticationModule()
+        {
+        }
+
+        // Security this method will access NetworkCredential properties that demand UnmanagedCode and Environment Permission
+        [EnvironmentPermission(SecurityAction.Assert, Unrestricted = true)]
+        [SecurityPermission(SecurityAction.Assert, Flags = SecurityPermissionFlag.UnmanagedCode)]
+        public Authorization Authenticate(string challenge, NetworkCredential credential, object sessionCookie, string spn, ChannelBinding channelBindingToken)
+        {
+            NetEventSource.Enter(this, "Authenticate");
```

Comment:
While I originally intended to do it this way, @geoffkizer made a good point in the logging PR that except for trivial cases (which this is), it becomes difficult for someone to know for sure that a logging call is not costing a significant amount, and so the easiest thing to do to be sure is to prefix all calls with ```if (NetEventSource.IsEnabled) ```.